Ability to read XDMF files and load data on the fly.

sebbelese opened the pull request:
Hi,

Please consider the following pull request.

With the included changes, the simulation results are loaded on the fly, separately for each time step and output field. Loaded results are kept in memory, and deleted only if used memory is above a certain limit, such that they do not need to be reloaded if we go back to an already loaded step or field. 

Only the core c++ lib has been modified, except for a small modification to accept the XDMF file format. The idea is that, istead of accessing output data directly through an access to the data array, we call a function that checks if the data is loaded or not before providing the array. I tried on relatively large datasets and did not notice any significant overhead.

Here are some limitations:
- The XDMF reader is not as general as the XDMF format is. It can read the files originating from our solver (scalars and vectors), but may need adjustements for certain other files.
- The range (Zmin, Zmax)  is not static anymore but is computed on the already loaded data.  It seems to not be updated every time it should be (it is updated only when we switch from Basic to Advanced colorbar). There is an update to call somewhere in the python files, but I do not know exactly where. 
- There is a maximum of 2Go RAM allowed for the simultaneously loaded outputs. This value should be set as an option.
- The datafile readers other than XDMF need to be updated to support on the fly data loading. They are not broken as it, but still load the entire data in memory. It should be relatively easy to do the changes, as soon as it is possible to load a single output.
- I had recourse to a limited amount of const_casts. It is probably possible to get rid of those, but I do not know enough the structure of the code to do it without significant modifications in the code design.
- For time series plot, all the data associated with the field needs to be loaded. It would be ablo to improve this by loading only the elements corresponding to the right location in the hdf5 file, but this would make the data readers more complex and I am not sure that this is needed.

As you see, there are some limitations and maybe some modifications are needed before the changes are integrated, but the plugin now perfectly fits my needs, while I could not use it before.

Thanks

PeterPetrik commented:
Hi, thank you for your work. It looks very good.

I have some general comments: 

- You use tabs, but we use spaces for intent (easily seen in the diff page). There is scripts/format_code.sh script that can be run to fix formatting issues. 
- To be able to test the code (both manually and automatically), we need some example results in the new format. Can you please add some small example to qgis-crayfish-plugin/test/data/ and add the simple test to qgis-crayfish-plugin/test/testload.py? Ideally the example should have 5 MB at maximum to keep the repository small.
- There are few new compilation warnings present on Linux, but most annoying is ./corelib/crayfish_mesh.h:55:29: warning: integer overflow in expression [-Woverflow]
because this header is included in almost all files
-  probably it is worth to create new function from this snippet 
`if (size == 0) dataSet->mesh()->updater->update(this, index, dataSet->index);`
- I do not understand why `inline const QVector<float> &getValues() const` updates the data, but non-const version not.

It seems to me that the new file format support is not problem to add and can be merged as-it-is (after adding tests). However we need to think about results simulation on-fly, because we probably need to have time-step plots and range updates working correctly.

sebbelese commented:
Hi,

Thank you for considering to merge the changes. I updated the sources, with some comments:

- Sorry for the tabs: I am a (new) user of VS and did not noticed. I ran the format script, but most of the files (including some not modified by me) have been reformated, since they were not in the specified format for other reasons than space/tab. The changes are in separate commits, so if needed they can be reverted with no big conflict hopefully.

- I added sample files in test/data. However, I do not have the environment set up to run testload.py, so I did not modify it. I am on Windows unfortunately. Is it easy to setup the environment to run the tests?

- The inline `QVector<float> &getValues()` non-const version is only called by data readers (xdmf, dat,...). For readers supporting on-the-fly read, this function is called to access the data pointer when the output is updating. Hence update should not be called, as it is already updating. Maybe another name should be given to this function, such as GetDataRef ?

- I think that the range is correctly updated by the c++ lib (considering all the already loaded steps), but it is just a matter of updating it in the gui when new data is loaded. The time-steps plot work correctly. However, the data is loaded when the plot is produced rather than when loading the datafile initially. At the end, it is not slower. Note that for the existing formats which have not been modified for on-the-fly loading, the behaviour has not been altered.

PeterPetrik commented:
will be merged from PR #248
